vw/src/main/scala/com/microsoft/azure/synapse/ml/vw/VowpalWabbitGeneric.scala

```diff
@@ -0,0 +1,195 @@
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.azure.synapse.ml.vw
+
+import com.microsoft.azure.synapse.ml.codegen.Wrappable
+import com.microsoft.azure.synapse.ml.core.env.StreamUtilities
+import com.microsoft.azure.synapse.ml.logging.BasicLogging
+
+import org.apache.spark.ml.{ComplexParamsReadable, ComplexParamsWritable, Estimator, Model}
+import org.apache.spark.ml.param.{Param, ParamMap, Params}
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.sql.catalyst.encoders.RowEncoder
+import org.apache.spark.sql.{DataFrame, Dataset, Row, functions => F}
+import org.apache.spark.sql.types.{StringType, StructField, StructType}
+import org.vowpalwabbit.spark.VowpalWabbitNative
+
+trait HasInputCol extends Params {
+  val inputCol = new Param[String](this, "inputCol", "The name of the input column")
+  def setInputCol(value: String): this.type = set(inputCol, value)
+  def getInputCol: String = $(inputCol)
+}
+
+/**
+  * This Estimator supports driving VW using VW-style examples (e.g. 0 |a b c)
+  */
+class VowpalWabbitGeneric(override val uid: String) extends Estimator[VowpalWabbitGenericModel]
+  with VowpalWabbitBaseLearner
+  with HasInputCol
+  with BasicLogging {
+  logClass()
+
+  override protected lazy val pyInternalWrapper = true
+
+  def this() = this(Identifiable.randomUID("VowpalWabbitGeneric"))
+
+  setDefault(inputCol -> "input")
+
+  override def copy(extra: ParamMap): this.type = defaultCopy(extra)
+
+  protected def getInputColumns(): Seq[String] = Seq(getInputCol)
+
+  override protected def trainFromRows(schema: StructType,
+                                       inputRows: Iterator[Row],
+                                       ctx: TrainContext): Unit = {
+
+    val featureIdx = schema.fieldIndex(getInputCol)
+
+    // learn from all rows
+    for (row <- inputRows) {
+      // ingestion and learning is collapsed
+      ctx.learnTime.measure {
+        ctx.vw.learnFromString(row.getString(featureIdx))
+      }
+
+      // trigger inter-pass all reduce
+      0 until ctx.synchronizationSchedule.getAllReduceTriggerCount(row) foreach { _ => ctx.vw.endPass() }
+    }
+  }
+
+  def fit(dataset: org.apache.spark.sql.Dataset[_]): VowpalWabbitGenericModel = {
+    // TODO: copied from predictor. should we merge train in here?
+    copyValues(train(dataset)
+      .setParent(this))
+  }
+
+  def transformSchema(schema: StructType): StructType = {
+    // validate the input column is here
+    // TODO: is this the right way of doing it? (as in expecting an exception?)
+    schema.fieldIndex(getInputCol)
+
+    schema
+  }
+
+  protected def train(dataset: Dataset[_]): VowpalWabbitGenericModel = {
+    logTrain({
+      val model = new VowpalWabbitGenericModel(uid)
+        .setInputCol(getInputCol)
+
+      trainInternal(dataset.toDF, model)
+    })
+  }
+}
+
+class VowpalWabbitGenericProgressive(override val uid: String)
```

Comment:
What does the Progressive mean here?

Reply:
progressive validation. iterative learning process works like this

- new example arrives
- predict using current weights
- update weight using example
- return prediction prior to update

(also called 1-step ahead prediction)